core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java

```diff
@@ -562,34 +566,55 @@ public RelOptPredicateList getPredicates(Exchange exchange,
-   * {@code (1, 2, null), (1, 2, null), (5, 2, null)} are {@code 'y = 2'} and {@code 'z is null'}.
+   * <p>The predicates on {@code T (w, x, y, z)} with rows
+   * {@code (1, 2, 3, null), (1, 2, null, null), (5, 2, 3, null)} are
+   * {@code 'SEARCH($0, Sarg[1, 5])'},
+   * {@code '=($1, 2)'},
+   * {@code 'SEARCH($2, Sarg[3; NULL AS TRUE])'} and
+   * {@code '[IS NULL($3)'}.
    */
   public RelOptPredicateList getPredicates(Values values, RelMetadataQuery mq) {
     ImmutableList<ImmutableList<RexLiteral>> tuples = values.tuples;
     if (!tuples.isEmpty()) {
-      Set<Integer> constants = new HashSet<>();
-      IntStream.range(0, tuples.size()).boxed().forEach(constants::add);
-      List<RexLiteral> firstTuple = new ArrayList<>(tuples.get(0));
-      for (ImmutableList<RexLiteral> tuple : tuples) {
-        if (constants.isEmpty()) {
-          break;
-        }
-        for (int i = 0; i < tuple.size(); i++) {
-          if (!Objects.equals(tuple.get(i), firstTuple.get(i))) {
-            constants.remove(i);
-          }
+      List<RexLiteral> firstTuple = tuples.get(0);
+      List<TreeSet<RexLiteral>> valueList = new ArrayList<>();
+      for (int i = 0; i < firstTuple.size(); i++) {
+        valueList.add(
+            i, new TreeSet<>((o1, o2) ->
+            Objects.equals(o1.getValue(), o2.getValue()) ? 0 : 1));
+      }
+      for (int i = 0; i < tuples.size(); i++) {
+        List<RexLiteral> tuple = tuples.get(i);
+        for (int j = 0; j < tuple.size(); j++) {
+          RexLiteral rexLiteral = tuple.get(j);
+          valueList.get(j).add(rexLiteral);
         }
       }
       RexBuilder rexBuilder = values.getCluster().getRexBuilder();
       List<RexNode> predicates = new ArrayList<>();
-      for (int i = 0; i < firstTuple.size(); i++) {
-        if (constants.contains(i)) {
-          RexLiteral literal = firstTuple.get(i);
+      for (int i = 0; i < valueList.size(); i++) {
+        TreeSet<RexLiteral> rexLiteralSet = valueList.get(i);
+        if (rexLiteralSet.size() == 1) {
+          for (RexLiteral rexLiteral : rexLiteralSet) {
+            predicates.add(i,
+                eqConstant(values, rexBuilder, i, rexLiteral));
+          }
+        } else {
+          RexUnknownAs rexUnknownAs = RexUnknownAs.UNKNOWN;
+          RangeSet<Comparable> rangeSet = TreeRangeSet.create();
+          for (RexLiteral rexLiteral : rexLiteralSet) {
+            if (RexUtil.isNull(rexLiteral)) {
+              rexUnknownAs = RexUnknownAs.TRUE;
+              continue;
+            }
+            rangeSet.add(Range.singleton(requireNonNull(rexLiteral.getValue5())));
+          }
+          final Sarg sarg = Sarg.of(rexUnknownAs, rangeSet);
           predicates.add(
-              rexBuilder.makeCall(SqlStdOperatorTable.EQUALS,
-                  rexBuilder.makeInputRef(literal.getType(), i),
-                  literal));
+              i, rexBuilder.makeCall(SqlStdOperatorTable.SEARCH,
```

Comment:
If we want to generate different expressions for a single element `($1, 5) or ISNULL($1)`, I think it is needed.

Reply:
This part of the code that uses `makeCall` with `SEARCH` does not generate `=($1,5)` and `IS NULL($1)` predicates so I don't completely follow the reasoning.

Reply:
We need to iterate once to obtain all values and remove duplicates, and then we can decide whether to generate SARG based on the size of the set.

Reply:
I try to use the `RexBuilder#makeIn` method to simplify the code.
But for VALUES SQL:
```
values(1, 2, 3, null), (1, 2, null, null), (5, 2, 3, null)
```
It generates the following predicates is :
```
[=($1, 2), IS NULL($3), OR(=($2, 3), =($2, null:NULL)), SEARCH($0, Sarg[1, 5])]
```
 `OR(=($2, 3), =($2, null:NULL))` it is wrong.  It is a bug?

Reply:
For VALUES:
```
values(1, 2, 3)
```
Will generate:
```
[SEARCH($0, Sarg[1]), SEARCH($1, Sarg[2]), SEARCH($2, Sarg[3])]
```
According to the comment in Jira. It is unacceptable.

Reply:
If the above two problems are solved in `rexBuilder#makeIn`, we can use it directly.

Reply:
The result of `RexBuilder.makeIn($2, {3, NULL})`, i.e., `OR(=($2, 3), =($2, null:NULL))` is not wrong but it is not the predicate that we want to create here. We want a predicate that guarantees that column `$2` is either `3` or `null`.

The singleton sargs case should be resolved after CALCITE-6608 but we don't have to wait for the latter to be merged; it's an independent improvement.

Anyways, at this stage I understand that we cannot use `makeIn` as is. The code here duplicates a bit the code inside the `RexBuilder` but it does what it intends to do. Ideally, it would be nice to find a way to refactor the code but it's not fair to block this PR till this happens.

I am OK, with the code as it is right now so I am marking this as resolved.